alert-common/src/main/java/com/synopsys/integration/alert/common/channel/issuetracker/IssueHandler.java

```diff
@@ -0,0 +1,198 @@
+ */
+public abstract class IssueHandler<T> {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private IssueTrackerMessageParser issueTrackerMessageParser;
+
+    public IssueHandler(IssueTrackerMessageParser issueTrackerMessageParser) {
+        this.issueTrackerMessageParser = issueTrackerMessageParser;
+    }
+
+    public final IssueTrackerMessageResult createOrUpdateIssues(IssueConfig issueConfig, MessageContentGroup content) throws IntegrationException {
+        Set<String> issueKeys = new HashSet<>();
+        for (ProviderMessageContent messageContent : content.getSubContent()) {
+            Set<String> issueKeysForMessage = createOrUpdateIssuesPerComponent(messageContent, issueConfig);
+            issueKeys.addAll(issueKeysForMessage);
+        }
+
+        String statusMessage = createStatusMessage(issueKeys);
+        return new IssueTrackerMessageResult(statusMessage, issueKeys);
+    }
+
+    protected Set<String> updateIssueByTopLevelAction(IssueConfig issueConfig, String providerName, LinkableItem topic, LinkableItem nullableSubTopic, ItemOperation action) throws IntegrationException {
+        if (ItemOperation.DELETE == action) {
+            List<T> existingIssues = retrieveExistingIssues(issueConfig.getProjectKey(), providerName, topic, nullableSubTopic, null, null);
+            logger.debug("Attempting to resolve issues in the project {} for Provider: {}, Provider Project: {}[{}].", issueConfig.getProjectKey(), providerName, topic.getValue(), nullableSubTopic);
+            Set<T> updatedIssues = updateExistingIssues(existingIssues, issueConfig, providerName, topic.getName(), action, Set.of());
+            return updatedIssues
+                       .stream()
+                       .map(this::getIssueKey)
+                       .collect(Collectors.toSet());
+        } else {
+            logger.debug("The top level action was not a DELETE action so it will be ignored");
+        }
+        return Set.of();
+    }
+
+    protected Set<String> createOrUpdateIssuesByComponentGroup(IssueConfig issueConfig, String providerName, LinkableItem topic, LinkableItem nullableSubTopic, SetMap<String, ComponentItem> groupedComponentItems)
+        throws IntegrationException {
+        Set<String> issueKeys = new HashSet<>();
+        String projectName = issueConfig.getProjectName();
+
+        SetMap<String, String> missingTransitionToIssues = SetMap.createDefault();
+        for (Set<ComponentItem> componentItems : groupedComponentItems.values()) {
+            try {
+                ComponentItem arbitraryItem = componentItems
+                                                  .stream()
+                                                  .findAny()
+                                                  .orElseThrow(() -> new AlertException(String.format("No actionable component items were found. Provider: %s, Topic: %s, SubTopic: %s", providerName, topic, nullableSubTopic)));
```

Comment:
Does it make sense to throw this exception? Or should we log a warning and continue to the next group?

Reply:
This exception will never actually be thrown. But throwing an exception is better than returning null because of our SonarCloud rules.